USWDS-Site: Add banner accessibility tests page

amyleadem opened the pull request:
# Summary

Added accessibility tests page for the banner component.

> [!Important]
> We should update changelog dates before merge.

## Related issue

Closes #2642 

## Preview link

- [Banner accessibility tests page](https://federalist-ead78f8d-8948-417c-a957-c21ec5617a57.sites.pages.cloud.gov/preview/uswds/uswds-site/al-a11y-tests-banner/components/banner/accessibility-tests/)
- [Banner main component page](https://federalist-ead78f8d-8948-417c-a957-c21ec5617a57.sites.pages.cloud.gov/preview/uswds/uswds-site/al-a11y-tests-banner/components/banner/)

## Resources
- [Creating accessibility checklists process doc](https://docs.google.com/document/d/14WTCsdzgwpoz_Miztd66RmEGfwwSEUFCVN1DkSQITTE/edit) (Google docs 🔒)
- [Accessibility checklist spreadsheet](https://docs.google.com/spreadsheets/d/1fAJaObQHv7aIaeUpH4fTLmauhmugMtNCrbpZe4BJcPY/edit#gid=97481116) (Google sheets 🔒)


## Testing and review
1. Confirm that both the main component page and the accessibility tests page have the correct compliance tag at the top.
1. Confirm that the main component page links to the accessibility tests page in the side navigation.
1. Confirm that the main component page links to the accessibility tests page from the accessibility guidance section.
1. Confirm that the main component page shows the accessibility tests summary.
1. Check that accessibility tests page provides the correct counts for each test status type.
1. Confirm the test checklist summaries and data are accurate.  
1. Confirm no visual issues.
1. Confirm there is an appropriate changelog entry on both the main component page and the accessibility tests page.

amyleadem commented on _data/accessibility-tests/banner.yml:
```diff
@@ -0,0 +1,112 @@
+    test_type: keyboard
+    version_tested: 3.8.0
+    wcag_criterion: 2.1.2
+  - summary: Banner is the first interactive object after the main navigation.
```
Curious - is this an accessibility test? It feels more like it would fall under something unrelated to a11y like guidance/standards compliance instead.

amyleadem commented on _data/accessibility-tests/banner.yml:
```diff
@@ -0,0 +1,112 @@
+    test_type: screen_reader
+    version_tested: 3.8.0
+    wcag_criterion: 2.4.6
+  - summary: Screen reader announces collapsed state.
```
Curious if this summary should also mention the expanded state.
```suggestion
  - summary: Screen reader announces collapsed and expanded states.
```

sarah-sch replied:
@amyleadem - deferring to @alex-hull and @amycole501 here.

alex-hull replied:
@amyleadem Yes, I think it should mention the collapsed and expanded state. Since it is an accordion, I think having both makes sense.

amyleadem replied:
Updated in [7b73480](https://github.com/uswds/uswds-site/pull/2687/commits/7b73480644d73da37edec395ac31622152e4dab1)

amyleadem commented on _data/accessibility-tests/banner.yml:
```diff
@@ -0,0 +1,112 @@
+    wcag_criterion: 4.1.2
+  - summary: Decorative elements are labeled appropriately.
+    summary_additional: |
+       When you access and open the banner while using a screen reader,
+       you don't hear any information about the flag, blue government building, or green lock icon.
+       When you view the alt text for the flag, blue government building, and green icon lock,
+       you see that they have been marked as decorative.
```
I am wondering if we can remove the second sentence here. It involves looking at code, which we have typically avoided. Additionally, the empty alt tag should evident by the screen reader not acknowledging the icons (as described in the first sentence):
```suggestion
       When you access and open the banner while using a screen reader,
       you don't hear any information about the flag, blue government building, or green lock icon.
```

Alternatively, maybe we could give some more explicit information about what an alt tag marked as decorative looks like:
```suggestion
       When you access and open the banner while using a screen reader,
       you don't hear any information about the flag, blue government building, or green lock icon.
       When you view the alt text for the flag, blue government building, and green icon lock,
       you see that the alt tag has an empty value.
```

amyleadem reviewed (commented):
@alex-hull @amycole501 @sarah-sch @finekatie 

I tagged you all with some open questions below. Please review and let me know if you have any questions!

sarah-sch commented on _data/accessibility-tests/banner.yml:
```diff
@@ -0,0 +1,112 @@
+    test_type: screen_reader
+    version_tested: 3.8.0
+    wcag_criterion: 4.1.2
+  - summary: Decorative elements are labeled appropriately.
```
Yes, good idea. Thanks!

sarah-sch commented on _data/accessibility-tests/banner.yml:
```diff
@@ -0,0 +1,112 @@
+    version_tested: 3.8.0
+    wcag_criterion: 1.4.11
+# Zoom/screen magnification tests
+  - summary: Text retains functionality when screen is magnified.
+    summary_additional: |
+      When you zoom in to 200%, text in the banner remains visible and retains all of its functionality.
```
Yes, I think "text remains visible" works is @amycole501 and @alex-hull agree.

amycole501 replied:
The only thing we might say is "When you zoom in to 200%, all text remains visible and the button text retains functionality".

amyleadem replied:
Updated in [fd7120b](https://github.com/uswds/uswds-site/pull/2687/commits/fd7120b411d3f6a5e908bdcf3de1279fdbe50a74)

amycole501 commented on _data/accessibility-tests/banner.yml:
```diff
@@ -0,0 +1,112 @@
+    version_tested: 3.8.0
+    wcag_criterion: 4.1.2
+  - summary: Decorative elements are labeled appropriately.
+    summary_additional: |
+       When you access and open the banner while using a screen reader,
+       you don't hear any information about the flag, blue government building, or green lock icon.
+       When you view the alt text for the flag, blue government building, and green icon lock,
+       you see that they have been marked as decorative.
```
I think the second alternative option is best since some teams may not have a screen reader available for some reason. The option for them to view the alt text is a nice option.

amyleadem replied:
Updated to the second suggestion in [fd7120b](https://github.com/uswds/uswds-site/pull/2687/commits/fd7120b411d3f6a5e908bdcf3de1279fdbe50a74)

amycole501 commented on _data/accessibility-tests/banner.yml:
```diff
@@ -0,0 +1,112 @@
+    wcag_criterion: 3.2.4
+  - summary: Icons meet color contrast requirements.
+    summary_additional: |
+       When you open the banner and use ANDI or another color contrast analyzer to look at the hex numbers,
+       contrast between the icons and the background is at least 4:5:1.
```
Yes please add that, thanks!

amyleadem replied:
Updated to `3:1` in [fd7120b](https://github.com/uswds/uswds-site/pull/2687/commits/fd7120b411d3f6a5e908bdcf3de1279fdbe50a74)

amycole501 commented on _data/accessibility-tests/banner.yml:
```diff
@@ -0,0 +1,112 @@
+    test_type: keyboard
+    version_tested: 3.8.0
+    wcag_criterion: 2.1.2
+  - summary: Banner is the first interactive object after the main navigation.
+    summary_additional: |
+      When you navigate your site with a keyboard and skip navigation,
+      the "Here's how you know" banner button is the first focusable item on the page.
```
This is for focus order. https://www.w3.org/WAI/WCAG21/Understanding/focus-order.html

Also I think it needs to be reworded a bit:
The Banner button is the first interactive object after the "skip navigation" link.  When you navigate your site with a keyboard and bypass the "skip navigation" link, the “Here’s how you know” banner button is the first focusable item on the page.

amycole501 replied:
Additionally, I think this may NOT be the case if a site alert is implemented. I think the [site alert ](https://designsystem.digital.gov/components/site-alert/)might be read aloud first AFTER the skip nav link. So the focus order would go:

1. Skip to main content
2. Site alert (if there are focusable items in the alert). Site alerts would be announced first to AT users, however if marked as "alerts".
3. Banner "here's how you know" button

amyleadem replied:
@amycole501 I updated the summary to match your suggestion in [390328f](https://github.com/uswds/uswds-site/pull/2687/commits/390328fb7edbaeec041978c6135abef062303ce0). Do we need to re-write the summary on this to account for the site alert exception, or leave it as is? 

@sarah-sch Do you want to proof the updated summary here?

sarah-sch replied:
@amyleadem - the update looks good.

amycole501 reviewed (commented):
I left a suggestion for a wording change to the focus order check. I also approved your suggestion for the alt text additional prompt.

sarah-sch commented on _data/accessibility-tests/banner.yml:
```diff
@@ -0,0 +1,112 @@
+    summary_additional: |
+       When you first access the "Here's how you know" button while using a screen reader,
+       the screen reader announces that it is "collapsed."
+       When activated, the screen reader announces "expanded".
```
Move the period inside of the quotation marks here. ("expanded.")

amyleadem replied:
Updated in [fd7120b](https://github.com/uswds/uswds-site/pull/2687/commits/fd7120b411d3f6a5e908bdcf3de1279fdbe50a74)

sarah-sch commented on _data/accessibility-tests/banner.yml:
```diff
@@ -0,0 +1,112 @@
+  - summary: Screen reader announces lock icon.
+    summary_additional: |
+       When you open the banner with a screen reader,
+       the screen reader uses the words "lock icon" to announce the lock icon.
```
Realizing that I think we should specify "white lock icon inside the parentheses" at the end of this sentence--otherwise people might confuse it with the other lock icon. Or they might overlook it altogether because it's not as obvious as the other lock icon. I think that without more specificity this test might cause confusion. (@amyleadem)

amyleadem replied:
I updated the copy to say "lock icon inside the parentheses" in [fd7120b](https://github.com/uswds/uswds-site/pull/2687/commits/fd7120b411d3f6a5e908bdcf3de1279fdbe50a74). I did not use the word "white" because the [default implementation](https://federalist-3b6ba08e-0df4-44c9-ac73-6fc193b0e19c.sites.pages.cloud.gov/preview/uswds/uswds/develop/?path=/story/components-banner--default) of the component has a black icon. Our site uses a different color combination than the standard.

sarah-sch commented on _data/accessibility-tests/banner.yml:
```diff
@@ -0,0 +1,112 @@
+  - summary: Screen reader does not announce decorative icons.
+    summary_additional: |
+       When you access and open the banner while using a screen reader,
+       you don't hear any information about the flag, blue government building, or green lock icon.
```
@amyleadem, @amycole501, @alex-hull - Should we specify "green circled lock icon" at the end of this sentence? There are two lock icons, and without greater specificity we are asking people to rely solely on color to distinguish between the two.

amycole501 replied:
Yes at first I didn't realize there were two, either!

amyleadem replied:
Added "green circled lock icon" in [fd7120b](https://github.com/uswds/uswds-site/pull/2687/commits/fd7120b411d3f6a5e908bdcf3de1279fdbe50a74)

amyleadem commented on _data/accessibility-tests/banner.yml:
```diff
@@ -0,0 +1,112 @@
+    test_type: general
+    version_tested: 3.8.0
+    wcag_criterion: 3.2.4
+  - summary: Icons meet color contrast requirements.
```
Added in [fd7120b](https://github.com/uswds/uswds-site/pull/2687/commits/fd7120b411d3f6a5e908bdcf3de1279fdbe50a74)

amyleadem commented:
@sarah-sch @alex-hull @amycole501 This is ready for your re-review. I believe I have addressed all of your comments. We are still [missing a test status for 2.4.6](https://github.com/uswds/uswds-site/pull/2687#discussion_r1609021751). @alex-hull @amycole501, can you let me know what status I should put there?

mahoneycm commented on _data/accessibility-tests/banner.yml:
```diff
@@ -0,0 +1,112 @@
+    summary_additional: |
+      When you use a screen reader and access the banner,
+      the screen reader uses the word "button" to announce the "Here's how you know" button.
+    test_status:
```
Just flagging that the test_status missing is causing a test count and visual issue. This will be resolved once status is added.

alex-hull replied:
@amyleadem I just listened to it in JAWS, and I hear "button" and it's collapsed/expanded state. I think we can mark this one as pass.

amyleadem replied:
Cool. Updated to "pass" in [7b73480](https://github.com/uswds/uswds-site/pull/2687/commits/7b73480644d73da37edec395ac31622152e4dab1)

mahoneycm commented on _data/accessibility-tests/banner.yml:
```diff
@@ -0,0 +1,121 @@
+component_status: pass
+test_items:
+# General tests
+  - summary: Color is not the only method used to convey meaning in the banner.
+    summary_additional: When viewing the banner, you never have to rely on color alone to understand meaning.
+    test_status: pass
```
Two notes here

This is marked as conditional on the spreadsheet. I could see either being suitable for this test since the default is passing but it will be up to the user if they make any customizations.

But I'm also curious if this test is necessary for this component since it doesn't use color at all to convey meaning. The way I understand it is that if you use color to define something (i.e. Calling it "_green_ lock icon" instead of just "lock icon") it's not the only identifier. Since this doesn't really use any color to describe meaning, do we need this test?

amyleadem replied:
@amycole501 @alex-hull 
What do you think about @mahoneycm's comment?

amycole501 replied:
I think we could probably remove that check in the interest of streamlining the checks overall. I'm guessing not many people will adjust the colors or content to make the check necessary.

alex-hull replied:
I agree with Amy. But am also keeping in mind that some people (most likely not in the banner, and some people may even get rid of those icons in banner) will definitely use color in their icons to convey meaning. A red stop sign on screen could be easy for sighted people to understand, but without other context, would be difficult for those with visual disabilities.

amyleadem replied:
Update: After more discussion, the a11y team decided that this test should stay in because there is a chance users might make edits that could make this test necessary. Going to leave this test as-is. 

cc: @alex-hull and @amycole501

mahoneycm commented on _components/banner/banner.md:
```diff
@@ -11,16 +11,8 @@ redirect_from:
-  href: '#banner-package'
-- text: Latest updates
-  href: '#changelog'
+  - text: Banner accessibility tests
+    href: /components/banner/accessibility-tests/
```
Comparing to other `subnav`'s, it looks like we should remove this indent

```suggestion
- text: Banner accessibility tests
  href: /components/banner/accessibility-tests/
```

mahoneycm commented on _data/accessibility-tests/banner.yml:
```diff
@@ -0,0 +1,121 @@
+    test_type: general
+    version_tested: 3.8.0
+    wcag_criterion: 1.4.1
+  - summary: Text meets color contrast requirements.
+    summary_additional: |
+      When you open the banner and use ANDI or another color contrast analyzer to look at the hex numbers,
+      the contrast between the banner text and background color is at least 4.5:1.
```
Curious, for components with settings to customize colors, should we add a note that they'll need to test themselves if they customize colors? We do offer color contrast checks that will warn them if the colors aren't compliant so maybe we don't need to worry about noting here.

What do you think?

amyleadem replied:
I don't think it is necessary to add an additional note here. We explicitly state at the top of the page that each of these tests should be completed after being implemented in a project ("USWDS tests components in isolation. You need to test the banner component in the context of your own site to ensure compliance with Section 508 accessibility standards."), so users should be testing regardless. Any objections to moving forward as-is?

alex-hull replied:
I'm leaning more with, Amy. I think we can keep it as is (but I also see your POV as well, Charlie).

mahoneycm reviewed (commented):
Code-wise, this is looking pretty good! Left a couple questions for ya but once your existing questions are cleared up this should be good to go

alex-hull reviewed (approved):
I think the tests look. Approving the additions.

sarah-sch reviewed (approved):
LGTM

thisisdano commented on _data/changelogs/component-banner-accessibility.yml:
```diff
@@ -0,0 +1,10 @@
+type: component
+changelogURL:
+items:
+  - date: NNNN-NN-NN
```
```suggestion
  - date: 2024-06-07
```

thisisdano commented on _data/changelogs/component-banner.yml:
```diff
@@ -2,10 +2,15 @@ title: Banner
 type: component
 changelogURL:
 items:
+  - date: NNNN-NN-NN
```
```suggestion
  - date: 2024-06-07
```
